### ISSUE TITLE
Delegated Mailbox Permission Module [Powershell] idiosyncracies are not well documented

### ISSUE THREAD

**Issue by dundir** [MicrosoftDocs/office-docs-powershell#6327]
For Powershell using the Exchange Online V2 module:

https://docs.microsoft.com/en-us/powershell/module/exchange/set-mailboxfolderpermission?view=exchange-ps

For either module:
Add-MailboxFolderPermission -identity "<user>:\calendar" 
Set-MailboxFolderPermission -identity "<user>:\calendar"

Using flags other than -AccessRights {Editor,FullControl} with -SharingPermissionFlags "{Delegate,None}" will cause a Watson Dump for O365/Exchange Online (Currently).  

While the docs do mention that SharingPermissionFlags needs to have AccessRights of at least Editor this wasn't listed in Get-Help. It also doesn't mention the fact that the module doesn't handle this cleanly when other options are chosen  (i.e. Module doesn't validate flags), 

As an example  -AccessRights "Reviewer" with -SharingPermissionFlags "None" causes a Watson dump to be triggered (O365 tenant) ) while throwing a warning about Editor permissions.

Reviewer Access to Calendars with Delegate status is a commonly used feature where they can read the calendar, but not receive messages about meetings. Its not possible to set this up properly without Outlook or Impersonation.

In cases where too much access is granted initially through Outlook it would be nice if we could fix delegate correctly without needing Outlook or Impersonation for Reviewers.

Additionally, the meeting message rule doesn't seem to get updated when correcting delegates via powershell without first setting SharingPermissionFlags "None" and it doesn't look possible currently for access levels other than Editor/FullControl without Outlook or Impersonation.

For common fixes after a user initially delegated a higher level privilege than necessary, an intermediate step is needed to stop the meeting messages for reviewers if all delegates are not first removed in Outlook.
 -AccessRights "Editor" -SharingPermissionFlags "None" followed by waiting some time (to update the local meeting message rule, before removing and recreating the Reviewer permission without the delegate flag.

The timetable for the local message rule to be updated when working on an Exchange Cached Profile doesn't appear to be documented. It seems to be immediate with an Online Profile. 
Some limited testing suggested Exchange Cached Mode updates the delegate related permissions, meeting message rule, and delegate list only once when the local Outlook client is first opened where the changes were initiated via Exchange Online Powershell.

There are many Outlook-only delegate options which do not have corresponding options with powershell. It would be nice to have a document that clearly shows existing coverage. (i.e. The more granular options for delegates are only configurable through Outlook client GUI or Impersonation).

Users with permissions will not appear in the delegate list without the delegate flag. Setting the delegate flag via powershell also sets the meeting message rule (silently). Updating the message rule with "None" also removes the user from the delegate list (while silently allowing Editor access privileges). 

Set-MailboxFolderPermission also doesn't appear to modify existing permissions. It atomically deletes and then recreates them. 
(i.e. setting Delegate Flag with an Editor permission using Add-MailboxFolderPermission and then changing Editor to Reviewer (without changing, or using SharingFlags) using Set-MailboxFolderPermission module causes what looks like the same watson dump as the add module when setting AccessRights and SharingFlags. 

Reviewers with this Delegate flag are needed for user controlled management of a user's mailbox/calendar. It would be nice if the documentation made the current limitations clear.

**Comment by dariomws** [comment]
Hi @dundir, thank you for your feedback and help us to improve docs.microsoft.com.

There is too much information here and it is difficult to differentiate each of them, I will do my best.

> For Powershell using the Exchange Online V2 module:
> 
> https://docs.microsoft.com/en-us/powershell/module/exchange/set-mailboxfolderpermission?view=exchange-ps
> 
> For either module:
> Add-MailboxFolderPermission -identity ":\calendar"
> Set-MailboxFolderPermission -identity ":\calendar"
> 
> Using flags other than -AccessRights {Editor,FullControl} with -SharingPermissionFlags "{Delegate,None}" will cause a Watson Dump for O365/Exchange Online (Currently).
> 
> While the docs do mention that SharingPermissionFlags needs to have AccessRights of at least Editor this wasn't listed in Get-Help. It also doesn't mention the fact that the module doesn't handle this cleanly when other options are chosen (i.e. Module doesn't validate flags),
> 
> As an example -AccessRights "Reviewer" with -SharingPermissionFlags "None" causes a Watson dump to be triggered (O365 tenant) ) while throwing a warning about Editor permissions.

_Online information is authoritative therefore the command-line version of this cmdlet will be updated soon._

> Reviewer Access to Calendars with Delegate status is a commonly used feature where they can read the calendar, but not receive messages about meetings. Its not possible to set this up properly without Outlook or Impersonation.
> 
> In cases where too much access is granted initially through Outlook it would be nice if we could fix delegate correctly without needing Outlook or Impersonation for Reviewers.

_For feature requests, I need to kindly redirect you to this resource: https://office365.uservoice.com/. There, your suggestion will reach the product team. Here we are limited to documentation issues and improvements._

> Additionally, the meeting message rule doesn't seem to get updated when correcting delegates via powershell without first setting SharingPermissionFlags "None" and it doesn't look possible currently for access levels other than Editor/FullControl without Outlook or Impersonation.
> 
> For common fixes after a user initially delegated a higher level privilege than necessary, an intermediate step is needed to stop the meeting messages for reviewers if all delegates are not first removed in Outlook.
> -AccessRights "Editor" -SharingPermissionFlags "None" followed by waiting some time (to update the local meeting message rule, before removing and recreating the Reviewer permission without the delegate flag.
> 
> The timetable for the local message rule to be updated when working on an Exchange Cached Profile doesn't appear to be documented. It seems to be immediate with an Online Profile.
> Some limited testing suggested Exchange Cached Mode updates the delegate related permissions, meeting message rule, and delegate list only once when the local Outlook client is first opened where the changes were initiated via Exchange Online Powershell.

I'll check meeting message rule to see if there are something that we can update/improve in the article.
 
> There are many Outlook-only delegate options which do not have corresponding options with powershell. It would be nice to have a document that clearly shows existing coverage. (i.e. The more granular options for delegates are only configurable through Outlook client GUI or Impersonation).
> 
> Users with permissions will not appear in the delegate list without the delegate flag. Setting the delegate flag via powershell also sets the meeting message rule (silently). Updating the message rule with "None" also removes the user from the delegate list (while silently allowing Editor access privileges).
> 
> Set-MailboxFolderPermission also doesn't appear to modify existing permissions. It atomically deletes and then recreates them.
> (i.e. setting Delegate Flag with an Editor permission using Add-MailboxFolderPermission and then changing Editor to Reviewer (without changing, or using SharingFlags) using Set-MailboxFolderPermission module causes what looks like the same watson dump as the add module when setting AccessRights and SharingFlags.
> 
> Reviewers with this Delegate flag are needed for user controlled management of a user's mailbox/calendar. It would be nice if the documentation made the current limitations clear.

I think here we have different topics, I'll check them.

**Comment by dariomws** [comment]
> Additionally, the meeting message rule doesn't seem to get updated when correcting delegates via powershell without first setting SharingPermissionFlags "None" and it doesn't look possible currently for access levels other than Editor/FullControl without Outlook or Impersonation.
> 
> For common fixes after a user initially delegated a higher level privilege than necessary, an intermediate step is needed to stop the meeting messages for reviewers if all delegates are not first removed in Outlook.
> -AccessRights "Editor" -SharingPermissionFlags "None" followed by waiting some time (to update the local meeting message rule, before removing and recreating the Reviewer permission without the delegate flag.

I could update permissions from Editor/Delegate to Reviewer, and the meeting message rule was updated:
![image](https://user-images.githubusercontent.com/33589238/90421287-8dab9980-e08f-11ea-99ac-0d840cb81290.png)

**Comment by dariomws** [comment]
> There are many Outlook-only delegate options which do not have corresponding options with powershell. It would be nice to have a document that clearly shows existing coverage. (i.e. The more granular options for delegates are only configurable through Outlook client GUI or Impersonation).

Maybe I don't understand but if there are missing options in the cmdlet article please let me know so I can update it. Check the following article for more information about delegate options in Outlook:
https://support.microsoft.com/en-us/office/allow-someone-else-to-manage-your-mail-and-calendar-41c40c04-3bd1-4d22-963a-28eafec25926
If you are suggesting to create a new article to compare/detail the delegate options in Outlook and PS, I cannot do that, I have to elevate it.
 
> Users with permissions will not appear in the delegate list without the delegate flag. Setting the delegate flag via powershell also sets the meeting message rule (silently). Updating the message rule with "None" also removes the user from the delegate list (while silently allowing Editor access privileges).

I think I need more detail to update the article. Using OWA I can see the same list of permissions I can see in PS:
![image](https://user-images.githubusercontent.com/33589238/90423377-b7b28b00-e092-11ea-8fb5-a7d8199f7872.png)

The article has the following description about the delegate flag:
"The user is made a calendar delegate, which includes receiving meeting invites and responses. If there are no other delegates, this value will create the meeting message rule. If there are existing delegates, the user is added to the meeting message rule without changing how delegate messages are sent."
I think the information about the rule update is there.

None description detail that removes delegate permissions and updates the rule.

> Set-MailboxFolderPermission also doesn't appear to modify existing permissions. It atomically deletes and then recreates them.
> (i.e. setting Delegate Flag with an Editor permission using Add-MailboxFolderPermission and then changing Editor to Reviewer (without changing, or using SharingFlags) using Set-MailboxFolderPermission module causes what looks like the same watson dump as the add module when setting AccessRights and SharingFlags.

I think it's the example in my screenshot and I didn't get the watson dump.

Anyway, I'll do a new review of your comments to double-check if I can add something to the article. Thanks for taking out some time to open the issue.

**Comment by dundir** [comment]
Hi Dario,

Thank you for reviewing this.

I just retested this and can confirm the Set-MailboxFolderPermissions module is now working properly without causing a Watson dump. When I initially posted, the Set-MailboxFolderPermission with -AccessRights Reviewer when modifying any rule with the SharingPermissionFlags set would throw a watson dump. It now appears to be silently clearing the SharingPermissionFlags and setting Reviewer permissions properly.

> If you are suggesting a document that compares Outlook with powershell coverage...

Yes, this is what I was looking for, specifically a comparison between Outlook Desktop and Powershell (OWA also as a plus). 
As a first step, when feature parity hasn't yet been achieved between two pieces of software, its very useful for people working with these systems to be able to tell at a glance whether a feature is available and will work or isn't without needing to perform any testing.

The sharing options through OWA require impersonation (either be the user or have full control as the user). I didn't see any issues with features missing with OWA but then again, I didn't spend much time testing that portion of the system.

The only other item was the timetable regarding when changes made to an account using those modules are pushed to Outlook clients running in Exchange Cached Mode.

**Comment by dariomws** [comment]
> I just retested this and can confirm the Set-MailboxFolderPermissions module is now working properly without causing a Watson dump. When I initially posted, the Set-MailboxFolderPermission with -AccessRights Reviewer when modifying any rule with the SharingPermissionFlags set would throw a watson dump. It now appears to be silently clearing the SharingPermissionFlags and setting Reviewer permissions properly.

_@dundir good to know that it's working now. If you had the latest module version I'm not sure how it was fixed. Maybe temporary network/connection problem, not sure._

> Yes, this is what I was looking for, specifically a comparison between Outlook Desktop and Powershell (OWA also as a plus).
> As a first step, when feature parity hasn't yet been achieved between two pieces of software, its very useful for people working with these systems to be able to tell at a glance whether a feature is available and will work or isn't without needing to perform any testing.
> 
> The sharing options through OWA require impersonation (either be the user or have full control as the user). I didn't see any issues with features missing with OWA but then again, I didn't spend much time testing that portion of the system.

I'll elevate this request to see if this article can be created.
 
> The only other item was the timetable regarding when changes made to an account using those modules are pushed to Outlook clients running in Exchange Cached Mode.

I'll test and/or try to get more information about it.

Thanks!

**Comment by dariomws** [comment]
> The timetable for the local message rule to be updated when working on an Exchange Cached Profile doesn't appear to be documented. It seems to be immediate with an Online Profile.
> Some limited testing suggested Exchange Cached Mode updates the delegate related permissions, meeting message rule, and delegate list only once when the local Outlook client is first opened where the changes were initiated via Exchange Online Powershell.

I opened a mailbox (MeganB) in Outlook with Cached Mode configured, removed the only delegate (AlexW) with PS, and sent a meeting to MeganB:
- MeganB received the meeting, not AlexW

Added AlexW again as delegate using PS, sent a meeting to MeganB:
- AlexW received the meeting, not MeganB

Based on this testing, when I update delegates using PS it's impacted immediately in Outlook client. Maybe I'm missing more detail from you.

**Comment by dundir** [comment]
I had run two tests, the first was similar to yours the only difference was the account contained an additional delegate as a placeholder to test the meeting rule functionality. There are typically only problems with the edge-case of updating the meeting rules when they are not completely removed. Removing the only delegate normally removes all meeting rules. In the case of multiple delegates it historically tends to break.

The second test involved adding a delegate in EO (and checking whether the delegate showed up in the delegate list in an Exchange Cached client (setup similar to the first test with multiple delegates), and then removing a client and testing. 

Changes originally seemed to only be triggered client side by closing the Client and reopening it. This seems to have been fixed as changes were propagated almost immediately when I reran the tests.

**Comment by dariomws** [comment]
> Changes originally seemed to only be triggered client side by closing the Client and reopening it. This seems to have been fixed as changes were propagated almost immediately when I reran the tests.

Ok, so I understand that we experienced the same behaviour and there is no need to update the article. Anyway, I'm checking if we can create an article as you requested, to compare delegate options.

**Comment by dariomws** [comment]
@dundir we'll proceed to close this issue. Anyway, I'm working on a new article to compare delegate options as you suggested. I'll leave a comment here when it's ready.

Thanks again, for taking out some time to open the issue. I appreciate and encourage you to do the same in the future also.
